source/common/network/apple_dns_impl.cc

```diff
@@ -402,6 +209,69 @@ AppleDnsResolverImpl::PendingResolution::dnsServiceGetAddrInfo(DnsLookupFamily d
       this);
 }
 
+void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
+    DNSServiceFlags flags, uint32_t interface_index, DNSServiceErrorType error_code,
+    const char* hostname, const struct sockaddr* address, uint32_t ttl) {
+  ENVOY_LOG(debug,
+            "DNS for {} resolved with: flags={}[MoreComing={}, Add={}], interface_index={}, "
+            "error_code={}, hostname={}",
+            dns_name_, flags, flags & kDNSServiceFlagsMoreComing ? "yes" : "no",
+            flags & kDNSServiceFlagsAdd ? "yes" : "no", interface_index, error_code, hostname);
+
+  // Generic error handling.
+  if (error_code != kDNSServiceErr_NoError) {
+    // TODO(junr03): consider creating stats for known error types (timeout, refused connection,
+    // etc.). Currently a bit challenging because there is no scope access wired through. Current
+    // query gets a failure status
+
+    pending_cb_.status_ = ResolutionStatus::Failure;
+    pending_cb_.responses_.clear();
+
+    finishResolve();
+    // Note: Nothing can follow this call to flushPendingQueries due to deletion of this
+    // object upon resolution.
+    return;
+  }
+
+  // Only add this address to the list if kDNSServiceFlagsAdd is set. Callback targets are only
+  // additive.
```

Comment:
IIUI, the Apple API supports removal, right? Are we for sure handling this correctly? Will it always "add" things that have already been added? Do we have to track state somehow? Can you add more comments?

Reply:
Consolidating below: https://github.com/envoyproxy/envoy/pull/17226#discussion_r664736728